### ISSUE TITLE
The easiest way to add multiple loss functions (with or without multiple optimizers) to the model

### ISSUE THREAD

**Issue by Anwarvic** [facebookresearch/fairseq#4179]
First of all, I really appreciate the effort and time spent on creating an amazing framework like this one. I can see the art every time I dig deep into the code. Great work!!

Here is the scenario, I'm trying to apply **alternate training** with an encoder-decoder model where I train different parts of the model based on a different loss function. In my case, I'm creating two loss functions:

- `criterion`: for the model as a whole.
-  `decoder_criterion`: for just the decoder.

What is the best way to do so without changing too much in the code-base? You should also know that every criterion could have a separate optimzier as mentioned in the question title

Below is how I did it, but I really believe there is a much better way.

# My Naive Way

- Added a new argument (`decoder_criterion=None`) to the `__init__()`,  `train_step`, and `valid_step`methods of the `Trainer` class.
- Added an additional new argument (`decoder_optimizer=None`) to the `train_step` method of the `Trainer` class.

These changes get the job done. And my code is running since yesterday with no problem. However, these changes aren't a few lines here and there. There are a lot of related codes that I needed to take care of for my code to work. Actually, I counted them. There were 53 other changes. Also, this is not scalable. What if I have one more loss function that measures the performance of a different part of the model! Then, I will have to add an additional argument. Imagine adding two or three more. It's gonna be a chaos.

I believe I shouldn't have changed the original signature of the mentioned methods. I think all should be done in the `FairseqCriterion` class. But I don't know how!

Any help is much appreciated!

**Comment by freddy5566** [comment]
Hi,
Does anyone have this issue? Apparently, fairseq does not support adding external criterions for now.
I wonder if anyone surpasses this restriction.

Thanks in advance.
